### ISSUE TITLE
ObjectMixin methods cause E_USER_DEPRECATED - BC Break

### ISSUE THREAD

**Issue by dakujem** [nette/utils#164]
- bug report? yes
- feature request? no
- version: `2.5.1`

### Description
I have just updated from `2.4.9` to `2.5.1` and ended up with a crashing application, because `ObjectMixin` now triggers error. 

Updating from `v2.4` to `v2.5` should not break the application.

### Steps To Reproduce
```php
error_reporting(E_USER_DEPRECATED);
ObjectMixin::hasProperty(new class extends stdClass{}, 'foo');  //  E_USER_DEPRECATED
```
This is also true for multiple other methods within the `ObjectMixin` class.

### Note
This is not the first issue of similar style. Please consider semantic versioning for nette package realeases.

**Comment by JanTvrdik** [comment]
Emitting E_USER_DEPRECATED is not a BC break and does not crash application.

**Comment by dg** [comment]
It is intention, you should not use ObjectMixin or Object.

**Comment by dakujem** [comment]
By what definition is it not a backwards-compatibility break? Enlighten me.
Let me repeat my statement:
> Updating from v2.4 to v2.5 should not break the application.

In v2.4  there is no error when calling `ObjectMixin::hasProperty`. In v2.5, there is an error. You do not find this a **change in behaviour**? Really?

---

Well I understand very well that you do not wish users to use these methods any more, after adding new class to replace the previous functionality. I just want you to **consider conforming to semantic versioning** for future releases.

How this should have been managed:

1. mark methods as `@deprecated` (instead of triggering errors) in version `2.5`, introduce new class and methods
2. wait until you release a backward incompatible version, e.g. `3.0`
3. throw error or remove deprecated parts of the code in `3.0`

It's that simple.

---

Now, I would not report this issue if it was the first time this has happened with Nette. I'm not braging or trying to dehonest your good work. I wouldn't waste my time to do that. On the contrary, I want you to avoid causing nuisances like this. 
So please consider this a friendly advice for future releases.

**Comment by dg** [comment]
The code still works unchanged, it just triggers the warnings that can be turned off.

**Comment by dakujem** [comment]
> Nette 2.x does not use semantic versioning

... I've noticed.

**Will Nette `>= 3.x` ?**

P.S.: I understand how errors work in PHP, I just wasn't satisfied with the versioning. But, if you say versions `2.x` do not conform to semantic versioning, then I accept it.

**Comment by milo** [comment]
A little bit offtopic, but still... PHP itself uses `E_DEPRECATED` between minor updates (e.g. http://php.net/manual/en/migration56.deprecated.php).

**Comment by Majkl578** [comment]
Deprecation notice is not a BC break since the original behavior is retained, only emits a notice that it won't be supported in future.
Of course [the approach used by Symfony/Doctrine](https://symfony.com/doc/current/contributing/code/conventions.html#deprecations) is more developer friendly - the deprecation messages are opt-in by default, but either way is backward compatible as long as the behavior does not change.

**Comment by dakujem** [comment]
@milo  PHP does not follow semantic versioning to begin with. A simple example is PHP 7.2 adding new keywords, breaking every software that uses `Object` named class...

@Majkl578 :
>  is backward compatible as long as the behavior does not change

Well, that is exactly my point of view! BUT, I find emitting an error a change in behaviour. That's all.

Whatever you may think, the fact remains: I needed to rewrite a piece of code to make it work. (Reconfiguring the error reporting settings was not an option). So I created this issue for Nette authors to consider a friendlier approach to semantic versioning for future releases.

P.S.: I like the doctrine approach

**Comment by Majkl578** [comment]
It's not a change in behavior, you can simply ignore deprecation notices by setting:
`ini_set('error_reporting', ini_get('error_reporting') ^ E_USER_DEPRECATED);`.
Then everything behaves as before. You don't need to rewrite anything as long as you decide you want to ignore the deprecations.
The only difference between Nette's approach and Symfony's/Doctrine's approaches is the silencing by `@` which has better DX, but has downsides (users unaware of such behavior will never know about deprecated code).

**Comment by dakujem** [comment]
How about reading my approach I listed above? You simply **mark** methods `@deprecated` in annotations (or in documentation) leaving the **functionality completely unchanged**. Completely. Then you remove the deprecated functinality in the following MAJOR release. Any sensible developer will read changelog or take extra care when updating the major version. No need to be extra cautious when updating through minor or patch versions. That is the added benefit resulting in more reliable code, less aversion to doing updates from developers and happier community in general.

And the whole point here is, *I do not want to change ANYTHING in my existing app*, not error reporting, not a single line of code. Imagine a case where you only want to quickly use a 3rd party component from componette, that actually uses 2.5 features. Now what?

**Comment by dakujem** [comment]
And please do not repeat that error reporting level thingie. I am aware of how it works. The point remains.

One can not say "It's exactly same, except for ...", it would be the same as stating "I'm not a racist, except I hate <put_a_colour_here> people.". Such a statement just does not count :roll_eyes:

**Comment by Majkl578** [comment]
> You simply mark methods @deprecated in annotations

Should be done anyway.

>  Then you remove the deprecated functinality in the following MAJOR release.

Correct, nothing was removed in a minor.

> Any sensible developer will read changelog or take extra care when updating the major version.

This is brave statement. Most of the developers will simply upgrade and see if anything breaks or throws deprecations (i.e. by running tests).

> I do not want to change ANYTHING in my existing app, not error reporting

You don't have to, just disable deprecations in php.ini if you don't care (btw [PHP disables deprecations too by default](https://github.com/php/php-src/blob/master/php.ini-production#L109)). Absolutely no need to change application code. :)

**Comment by dakujem** [comment]
Okay, let me change that to

> I do not want to change ANYTHING in my existing environment

How about now?

**Comment by Majkl578** [comment]
> I do not want to change ANYTHING

And yet you've just updated your dependencies.

Deprecations are normal and are part of standard software development. They are also covered by semver itself: https://semver.org/#how-should-i-handle-deprecating-functionality
Adding deprecations in minors is perfectly legit and semver compliant (Nette does that, Symfony does that, Doctrine does that, everyone does).
There is no reason to discuss this any further.

**Comment by milo** [comment]
@dakujem 
> I do not want to change ANYTHING in my existing app
So, don't install any 3rd party lib. Or install 3rd party lib for Nette 2.4.

You are willing to edit your application, update app dependencies, but you don't want change one single easy thing in PHP behaviour.

> I do not want to change ANYTHING in my existing environment
In that case your env setting is buggy. Fix it and don't blame the deprecated messages.

How is the `E_USER_DEPRECATED` message BC break? If it kills your app in production, your app is buggy.

**Comment by dakujem** [comment]
Semver org is the resource I follow when upgrading packages.

It's your interpretation of (probably) the following line that bothers me (copied from semver.org): 
> issue a new minor release with the deprecation in place.

But hey! :bulb: I get your point. You want to FORCE users to stop using the deprecated parts BEFORE they upgrade to `3.0`. Well your choice. I would do it as I described in one of y earlier posts.

Yet you guys don't get it at all. My message to you was to make the deprecations more friendly, nothing else. No harm in that. So why do you keep braging about it? Furthermore, if David says Nette 2.x does not follow semantic versioning, than there really is no point in this debate.

Unless Nette `3.x` and following releases are made to follow semantic veriosning. And I hope they are. I'm willing to help there.

**Comment by milo** [comment]
Hm, probably understood. I think that merits of the discussion came from the issue description: "It is a BC break because it crashed your application." Personally, I can't find any way how `E_USER_DEPRECATED` can crash the application.

Imagine, you had `E_DEPRECATED` notification disabled before your update to 2.5. You will never notice about this change. Code will work without any change. There is no BC break. It is all about logging only. And development mode is about logging to maximum. But you are not forced to use it, it is only default behaviour. Just turn it off during development and you are fine too. No needs to change the app code.

**Comment by dakujem** [comment]
Well, yes, you're right with the logging setting, but that could be said for notices and most other errors too. Though that would be the developer's fault anyway. I understand that this is a different case as the triggerring of `E_USER_DEPRECATED` here is intentional.

We're using very strict settings and do get all the errors. This error was obviously caught during dev/testing phase. It irritated me, because I approach deprecations differently (as described somewhere above) and I dislike having to repair code right after doing an update or two (considering it should have been just minor feature update).

This is obviously a grey zone in PHP. One of many, unfortunatelly. I can't really blame you for the bug because our configuration might differ from what you intended the lib for. On the orher hand, you can't really say I'm completely wrong either, because such a configuration does indeed exist.

Thanks for resolving this.

**Comment by dg** [comment]
I think Nette makes the deprecations friendly. The deprectated API is silently marked as `@deprecated` in one version, in the next one it triggers warnings and finally in the third version is removed. 

By the way, it exactly matches the **How should I handle deprecating functionality?** on semver.org.

However, this is special issue. Using of Nette\Object in PHP 7.2 will cause a fatal error that lacks the stack trace, so sometimes it is very hard to discover where the Nette\Object is still used. So I released v2.5 only to make deprecation even more friendly.

**Comment by dakujem** [comment]
Thank you for clarification, David.

I agree on the `Nette\Object` matter in the way how you solved it in Nette packages (by replacing it with traits etc.). We have some code written on top of nette/utils `ObjectMixin` so that is where the problem arised.
